### ISSUE TITLE
Lots of "could not parse assembly" error messages

### ISSUE THREAD

**Issue by trask** [AdoptOpenJDK/jitwatch#159]
Jitwatch is logging lots of "could not parse assembly" error messages.  I haven't noticed it affecting my usage of Jitwatch, but wanted to report the details in case it's useful.

In a large (2.9gb) hotspot log I analyzed, it logged

* 386,000 similar to `Could not parse assembly: OopMap{[0]=Oop [112]=Oop [120]=Oop [128]=Oop [136]=Oop [144]=Oop off=10728}`
* 20,000 similar to `Could not parse assembly: OopMapSet contains 0 OopMaps`
* 15,000 similar to `Could not parse assembly: Compiled method (c1)    2319   99       2       java.lang.Character::toLowerCase (9 bytes)`
* 4,000 similar to `Could not parse assembly: Compiled method (c2)    2322   51       4       java.lang.String::startsWith (7 bytes)`
* 100 similar to `Could not parse assembly: Compiled method (nm)    4191  495     n 0       java.lang.String::intern (native)`
* 50 similar to `Could not parse assembly: Decoding VtableStub vtbl[104]@12`
* 30 similar to `Could not parse assembly: Decoding VtableStub itbl[10]@12`
* 20 similar to `Could not parse assembly: argument handler *#*103 for: static java.lang.reflect.Array.set(Ljava/lang/Object;ILjava/lang/Object;)V (fingerprint = 1387325, 50 bytes generated)`
* and a handful of others

The hotspot log was generated on Linux/8u40.

**Comment by chriswhocodes** [comment]
Hi,

I believe this is caused by the log mangling that occurs when the HotSpot LogCompilation (XML), TraceClassLoading, and PrintAssembly writer threads clash when writing the log file.

Could you pick one of those assembly parse errors and paste 10 lines of context each side and I'll see if there's any way I can make my parser more intelligent?

Cheers,

Chris

**Comment by trask** [comment]
Hey, so first, sorry, I was running JMH perfasm profiler at the same time which enabled -XX:+PrintNMethods and caused most of the parse assembly errors.  After removing the perfasm profiler, the number of parse assembly errors was down to a much more reasonable 1069.  And they were all like you said due to assembly lines getting broken in two by the other loggers.  I submitted PR #160 that makes a few fallback attempts when it encounters assembly parse errors.  This cleaned up all of the 1069 assembly errors in the hotspot log that I was loading.  The conditionals kinda exploded, so if you'd prefer, I'll update the PR and re-write the parseAssembly method using a stateful parsing object so the loop body can be decomposed into a bunch of smaller methods.  Or any other suggestions just let me know.  Oh yeah, I'll add some tests for the various assembly line breakage I ran into.  I probably should have done that first :-).  Thanks.
